[RF-DOCS] Rails Routing Guide [ci-skip] (#52521)
rails/rails: guides/source/routing.md

The Purpose of the Rails Router
-------------------------------

The Rails router matches incoming HTTP requests to specific controller actions
in your Rails application based on the URL path. (It can also forward to a
[Rack](rails_on_rack.html) application.) The router also generates path and URL
helpers based on the resources configured in the router.

### Routing Incoming URLs to Code

> When your Rails application receives an incoming request, it asks the router to match it to a controller action. For example, take the following incoming request:

Reviewer: We can match a route to [any rack app](https://guides.rubyonrails.org/routing.html#routing-to-rack-applications), which not necessary is a controller action. I'm not sure which wording to use, though.

Author: edited. Yes we do mention Rack in the intro, right above.

```
GET /users/17
```

If the first matching route is:

```ruby
get '/users/:id', to: 'user#show'
```

> The request is matched to the `UsersController` class's `show` method with `{ id: '17' }` in the `params` hash.

Author: haha, changing it to "method" was requested by another reviewer. Here's what I'll do: change it back to action except up top where I say action (aka method).

> The request is matched to the `UsersController` class's `show` method with `{ id: '17' }` in the `params` hash.

Reviewer 2: Ha 😄 . Sometimes I worry I'll make a suggestion that reverses a suggestion I made earlier.

> The request is matched to the `UsersController` class's `show` action with `{ id: '17' }` in the `params` hash.

Reviewer 3: Since we change the naming of `controller` to `class` I wonder if we shouldn't change also `action` to `method` since that's in same paradigm.

Reviewer 4: Suggestion to reword: "to the `UsersController` class's `show` action" -> "to the `show` action in `users` controller with...". This keeps it in the Rails paradigm.

The `to:` option expects a `controller#action` format when passed a string. Alternatively, You can pass a symbol and use the `action:` option, instead of `to:`. You can also pass a string without a `#`, in which case the `controller:` option is used instead to `to:`. For example:

```ruby

> get '/users/:id', action: :show, controller: 'users'

Reviewer: How about passing the controller first, then the action? I think keeping the same order as `controller#action` helps people to visualize what changed.
```suggestion
get '/users/:id', controller: 'users', action: :show
```

```

> NOTE: Rails uses snake_case for controller names when specifying routes. For example, if you have a controller named UserProfilesController, you would specify a route to the show action as user_profiles#show.

Reviewer 3: ```suggestion
NOTE: Rails uses snake_case for controller names when specifying routes. For example, if you have a controller named `UserProfilesController`, you would specify a route to the show action as `user_profiles#show`.
```

### Generating Paths and URLs from Code

> The Router automatically generates paths and URLs that can be used throughout your application, instead of hard-coded strings.

Author: Will go with the 2nd one. The detail about the :as option might be too early here.

For example, `user_path` and `user_url` helpers in the route below.

```ruby
get '/users/:id', to: 'users#show', as: 'user'
```

NOTE: The `as:` option is used to provide a custom name for a route, which is used when generating URL and path helpers.

Assuming your application contains this code in the controller:

```ruby
@user = User.find(params[:id])
```

and this in the corresponding view:

```erb
<%= link_to 'User Record', user_path(@user) %>
```

> The router will generate the path `/users/17` from `user_path(@user)`. Using the `user_path` helper allows you to avoid having to hard-code a path in your views.

Reviewer: Does it make sense to explain _why_ not using hardcoded strings is helpful?
```suggestion
The router will generate the path `/users/17` from `user_path(@user)`. Using the `user_path` helper allows you to avoid having to hard-code a path in your views. This is helpful if you eventually move the route to a different URL, as you won't need to update the corresponding views.
```

Author: Sure!

> It also generates `user_url`, which has a similar purpose. While `user_path` generates a relative url like `/users/17`, `user_url` generates an absolute url something like `https://yourdomain.com/users/17` in the above example.

Reviewer 5: Could we use `example.com` instead of `yourdomain.com` here and in the other instances in this PR?

end
```

> Since this is a regular Ruby source file you can use all of Ruby's features to help you define your routes.

Reviewer: What about?
```suggestion
Since this is a regular Ruby source file, you can use all of Ruby's features (like conditionals and loops) to help you define your routes.
```

NOTE: The `Rails.application.routes.draw do ... end` block that wraps your route definitions is required to establish the scope for the router DSL (Domain Specific Language) and must not be deleted.

NOTE: Be careful with variable names in `routes.rb` as they can clash with the DSL methods of the router.

Resource Routing: the Rails Default
-----------------------------------

Resource routing allows you to quickly declare all of the common routes for a given resource controller. For example, a single call to [`resources`][] declares all of the necessary routes for the `index`, `show`, `new`, `edit`, `create`, `update`, and `destroy` actions, without you having to declare each route separately.

[`resources`]: https://api.rubyonrails.org/classes/ActionDispatch/Routing/Mapper/Resources.html#method-i-resources

### Resources on the Web

> Browsers request pages from Rails by making a request for a URL using a specific HTTP verb, such as `GET`, `POST`, `PATCH`, `PUT`, and `DELETE`. Each HTTP verb is a request to perform an operation on the resource. A resource route maps related requests to methods (aka actions) in a single controller.

Reviewer 4: I'd consider just using actions and removing single here. So:

```suggestion
Browsers request pages from Rails by making a request for a URL using a specific HTTP verb, such as `GET`, `POST`, `PATCH`, `PUT`, and `DELETE`. Each HTTP verb is a request to perform an operation on the resource. A resource route maps related requests to actions in a controller.
```

Author: edited

| PATCH/PUT | /photos/:id      | photos#update     | update a specific photo                      |
| DELETE    | /photos/:id      | photos#destroy    | delete a specific photo                      |

Since the router uses the HTTP verb *and* path to match inbound requests, four URLs can map to seven different controller actions. For example, the same `photos/` path matches to `photos#index` when the verb is `GET` and `photos#create` when the verb is `POST`.

NOTE: Order matters in the `routes.rb` file. Rails routes are matched in the order they are specified. For example, if you have a `resources :photos` above a `get 'photos/poll'` the `show` action's route for the `resources` line will be matched before the `get` line. If you want the `photos/poll` route to match first, you'll need to move the `get` line **above** the `resources` line.

### Path and URL Helpers

> Creating a resourceful route will also expose a number of helpers to controllers and views in your application.

Author: Good idea. Added to the TIP below.

### Singular Resources

Sometimes, you have a resource that users expect to have only one (i.e. it does not make sense to have an `index` action to list all values of that resource). In that case, you can use `resource` (singular) instead of `resources`.

The below resourceful route creates six routes in your application, all mapping to the `Geocoders` controller:

```ruby
resource :geocoder
resolve('Geocoder') { [:geocoder] }
```

> NOTE: The call to `resolve` is necessary for converting instances of the `Geocoder` to singular routes through [record identification](form_helpers.html#relying-on-record-identification).

Author: Hm...not clear on this suggestion, maybe you can show is using "Add a suggestion"? (the NOTE is referring to the 'resolve' right above, might be confusing to move it)

resources :articles, shallow: true do
  resources :comments
  resources :quotes
end
```

> The articles resource above will generate the following routes:

Author: Open the 2nd one, don't feel strongly. Though it is accurate as is since `articles` (plural) is by rails convention and while the method in the routes files is `resources`, this sentence is referring to the word "resource" and not the method per se. But open to updating to the 2nd one is others feel strongly, fine either way.

end
```

> An incoming GET request to `/photos/1/preview` will route to the `preview` action of `PhotosController`. The resource id value will be available in `params[:id]` . It will also create the `preview_photo_url` and `preview_photo_path` helpers.

Reviewer 6: Extra space after `params[:id]`

Reviewer: ```suggestion
An incoming GET request to `/photos/1/preview` will route to the `preview` action of `PhotosController`. The resource id value will be available in `params[:id]`. It will also create the `preview_photo_url` and `preview_photo_path` helpers.
```

TIP: If you find yourself adding many extra actions to a resourceful route, it's time to stop and ask yourself whether you're disguising the presence of another resource.

It is possible to customize the default routes and helpers generated by `resources`, see [customizing resourceful routes section](#customizing-resourceful-routes) for more.

Non-Resourceful Routes

> ----------------------

Reviewer 2: This section has a lot of subsection, some which are pretty short compared to other guides.
Could we maybe group some of these and use a smaller heading for the grouped sections?
For example group the segments (I've added more sections here) and constraints sections?
Maybe something like:
- Routing Segments
  - Bound Parameters
  - Dynamic Segments
  - Static Segments
  - Wildcard Segments
  - Format Segments
  - The Query String
  - Defining Default Parameters
  - Naming Routes
  - Unicode Character Routes
- Routing Contraints
  - HTTP Verb Constraints
  - Segment Constraints
  - Request-Based Constraints
  - Advanced Constraints
- Redirection
- Using root
- Routing to Rack Applications
- Direct Routes
- Using resolve

Author: Hi @p8, this section does have a lot of sub-sections and we could group them. I think the next heading level h4 won't show up in the nice fancy scrolling index thing on the right? Surfacing these "features" there feels helpful to the readers in discovering the content. Otherwise they'd have to just scroll around to find it. My sense is "Segments" and "Constraints" headings aren't revealing enough unless you already know a lot about routing.  What do you think? I'm open to grouping it like you showed above. But thought I'd mention the index thing.

Reviewer 2: I don't have a strong opinion on this. Let's keep it as is.

In addition to resourceful routing with `resources`, Rails has powerful support for routing arbitrary URLs to actions. You don't get groups of routes automatically generated by resourceful routing. Instead, you set up each route separately within your application.

> While you should usually use resourceful routing, there are places where non-resourceful routing is more appropriate. There's no need to try to shoehorn every last piece of your application into a resourceful framework if that's not a good fit.

Reviewer: I wonder if we should use a simpler language here. While the term fits perfectly, many non-native English speakers won't know it (I didn't!).
```suggestion
While you should usually use resourceful routing, there are places where non-resourceful routing is more appropriate. There's no need to try to force every last piece of your application into a resourceful framework if that's not a good fit.
```

get 'photos(/:id)', to: 'photos#display'
```

> If an incoming request of `/photos/1` is processed by this route, then the result will be to invoke the `display` action of the `PhotosController`, and to make the final parameter `"1"` available as `params[:id]`. This route will also route the incoming request of `/photos` to `PhotosController#display`, since `:id` is an optional parameter, denoted by parentheses in the above example.

Reviewer: ```suggestion
If an incoming `GET` request of `/photos/1` is processed by this route, then the result will be to invoke the `display` action of the `PhotosController`, and to make the final parameter `"1"` available as `params[:id]`. This route will also route the incoming request of `/photos` to `PhotosController#display`, since `:id` is an optional parameter, denoted by parentheses in the above example.
```

Reviewer 2: Should this also be `incoming `GET` request for`?
```suggestion
If an incoming `GET` request for `/photos/1` is processed by this route, then the result will be to invoke the `display` action of the `PhotosController`, and to make the final parameter `"1"` available as `params[:id]`. This route will also route the incoming request of `/photos` to `PhotosController#display`, since `:id` is an optional parameter, denoted by parentheses in the above example.
```

get 'photos/:id/:user_id', to: 'photos#show'
```

> This route will respond to paths such as `/photos/1/2`. The `params` hash will be  { controller: 'photos', action: 'show', id: '1', user_id: '2' }.

Reviewer 6: will be `{ controller: ...}`.

get 'photos/:id', to: 'photos#show'
```

> An incoming `GET` request for `/photos/1?user_id=2` will be dispatched to the `show` action of the `PhotosController` class as usual and the `params` hash will be `{ controller: 'photos', action: 'show', id: '1', user_id: '2' }`.

Reviewer: I _think_ both this section and the next one are also applicable to resourceful routes. This is not a must-have for this PR, but do you have any ideas for a way to make it clear that you can use this in resourceful routes as well?

Author: I was thinking that, I'll make a follow-on task to review and test all of these features in this section. More than one are likely applicable to resourceful routes...and things can be restructured a bit to make that clear.

```ruby
defaults format: :json do
  resources :photos

>   resources :articles

Reviewer 6: Do you need to include it here?

Author: include what?

```ruby
get 'exit', to: 'sessions#destroy', as: :logout
```

This will create `logout_path` and `logout_url` as the route helpers in your application. Calling `logout_path` will return `/exit`.

You can also use `as` to override routing helper names defined by `resources` by placing a custom route definition *before* the resource is defined, like this:

```ruby
get ':username', to: 'users#show', as: :user
resources :users
```

This will define a `user_path` helper that will match `/:username` (e.g. `/jane`). Inside the `show` action of `UsersController`, `params[:username]` will contain the username for the user.

### HTTP Verb Constraints

In general, you should use the [`get`][], [`post`][], [`put`][], [`patch`][], and [`delete`][] methods to constrain a route to a particular verb. There is a [`match`][] method that you could use with the `:via` option to match multiple verbs at once:

```ruby
match 'photos', to: 'photos#show', via: [:get, :post]
```

The above route matches GET and POST requests to the `show` action of the `PhotosController`.

You can match all verbs to a particular route using `via: :all`:

```ruby
match 'photos', to: 'photos#show', via: :all
```

> NOTE: Routing both `GET` and `POST` requests to a single action has security implications. In general, you should avoid routing all verbs to an action unless you have a good reason.

Reviewer: Is the next NOTE one of the "security implications" of this? If so, could it be just a single NOTE?

Author: Sure!

get 'photos/:id', to: 'photos#show', constraints: { id: /[A-Z]\d{5}/ }
```

> The above route definition requires `id` to be 5 alphanumeric characters long. Therefore, this route would match paths such as `/photos/A12345`, but not `/photos/893`. You can more succinctly express the same route this way:

Reviewer: The guide states that `:contraints` takes a regular expression, but I wonder if that's 100% correct. I think it accepts any object that responds to `#matches?` or `#call`

Author: Yes. We cover that in the "Advance Constraints" section below. It can take custom class that implements `matches?`, lambda, etc.

I added a note to this section.

end
```

### Wildcard Segments

> A route definition can have a wildcard segment, which is a segment prefixed with a star, such as`*other`:

Reviewer 6: Missing space next to _as `*other`_.

Reviewer: ```suggestion
A route definition can have a wildcard segment, which is a segment prefixed with a star, such as `*other`:
```

would match `books/some/section/last-words-a-memoir` with `params[:section]` equals `'some/section'`, and `params[:title]` equals `'last-words-a-memoir'`.

Technically, a route can have even more than one wildcard segment. The matcher assigns segments to parameters in the order they occur. For example:

```ruby
get '*a/foo/*b', to: 'test#index'
```

would match `zoo/woo/foo/bar/baz` with `params[:a]` equals `'zoo/woo'`, and `params[:b]` equals `'bar/baz'`.

### Format Segments

Given this route definition:

```ruby
get '*pages', to: 'pages#show'
```

By requesting `'/foo/bar.json'`, your `params[:pages]` will be equal to `'foo/bar'` with the request format of JSON. If you want to match URLs without an explicit format and ignore URLs that include a format extension, you could supply `format: false` like this:

```ruby
get '*pages', to: 'pages#show', format: false
```

> If you want to make the format segment mandatory, so it cannot be omitted, you can supply `format: true` like this:

Reviewer: Is this the default behavior? Is there a default behavior?

Author: Added "The default behavior with `format` is that if included Rails automatically captures it from the URL and includes it in params[:format], but `format` is not required in a URL."

root 'pages#main' # shortcut for the above
```

You typically put the `root` route at the top of the file so that it can be matched first.

> NOTE: The `root` route only routes `GET` requests to the action.

Reviewer 6: I'm afraid this is not true, e.g:

```ruby
  root "posts#index", via: :post
```

Generates:

```shell
Prefix Verb URI Pattern Controller#Action
  root POST /           posts#index
```

Author: Good catch. Updated!

| PATCH/PUT | /photos/:id      | photos#update     | image_path(:id)      |
| DELETE    | /photos/:id      | photos#destroy    | image_path(:id)      |

> ### Renaming the `new` and `edit` Helper Prefix

Reviewer: I think so. I thought `/resource/:id/delete` existed, but I think I'm wrong.

| PATCH/PUT | /photos/:id      | photos#update     | image_path(:id)      |
| DELETE    | /photos/:id      | photos#destroy    | image_path(:id)      |

> ### Renaming the `new` and `edit` Helper Prefix

Reviewer: I think this section might be wrong. As far as I could test, this _doesn't_ change the helper prefix, only the path segments. Can you double-check?

Author: You're right, good catch! That's too bad as the other way would be more consistent.

| PATCH/PUT | /photos/:id      | photos#update     | image_path(:id)      |
| DELETE    | /photos/:id      | photos#destroy    | image_path(:id)      |

### Renaming the `new` and `edit` Helper Prefix

> The `:path_names` option lets you override the default `new` and `edit` prefix in path helpers. For example:

Reviewer: The title this sentence still claim that this changes helper prefixes. We should change that to "segment in paths" again

form_with(model: [@account, @article])   # => <form action="/1/article/9" ...>
```

> The `:as` option is also not required, but without it, Rails will raise an error when evaluating `url_for([@account, @article])` or other helpers that rely on `url_for`, such as [`form_with`][].

Reviewer: ```suggestion
The `:as` option is also not mandatory, but without it, Rails will raise an error when evaluating `url_for([@account, @article])` or other helpers that rely on `url_for`, such as [`form_with`][].
```

[`form_with`]: https://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html#method-i-form_with

### Restricting the Routes Created

> By default, using `resources` creates routes for the seven default actions (`index`, `show`, `new`, `create`, `edit`, `update`, and `destroy`). You can use the `:only` and `:except` options to limit which routes are created.

Reviewer: GitHub doesn't allow me to comment on unchanged lines, but could you update the line breaks at lines 1212-1213? They look strange in the rendered version

<img width="891" alt="image" src="https://github.com/user-attachments/assets/be6b9c38-42cd-45f4-aadd-682a94decc02">

The `:only` option tells Rails to create only the specified routes:

```ruby
resources :photos, only: [:index, :show]
```

> Now, a `GET` request to `/photos` or `/photos/:id` would succeed, but a `POST` request to `/photos` will fail to match.

Reviewer: We could mention the `routes --unused` command at line 1275.

Author: Added to the TIP

| PATCH/PUT | /kategorien/:id            | categories#update  | category_path(:id)      |
| DELETE    | /kategorien/:id            | categories#destroy | category_path(:id)      |

> ### Overriding the Singular Form

Reviewer: I might have missed this elsewhere, but is there any reason why this was removed? I find it useful to have at least a tip or link mentioning custom inflections.

Author: Hm...I *think* we felt that inflections aren't to the Routing and maybe a bit extraneous. But open to adding back in.

Reviewer: I guess a tip would be helpful: "if you need to override how routes are pluralized, check our inflector docs" or something along those lines.

Reviewer: This can be a separate PR, not to block this one.

Author: Added it back (and reworded slightly). If you find it useful, someone else might as well :)

> ### Using `:as` in Nested Resources

Reviewer: I wonder if this should be part of `Overriding the Named Route Helpers`

Author: I think it works here too as the nested example is a bit more complicated and having it in two places reinforces the information. Resolving for now, but let me know if you feel strongly, open to restructuring as well.

The `:as` option can override routing helper names for resources in nested routes as well. For example:

```ruby
resources :magazines do
  resources :ads, as: 'periodical_ads'
end
```

This will create routing helpers such as `magazine_periodical_ads_url` and `edit_magazine_periodical_ad_path` instead of the default `magazine_ads_url` and `edit_magazine_ad_path`.

### Renaming Default Route Parameter `id`

It is possible to rename the default parameter name `id` with the `:param` option. For example:

```ruby
resources :videos, param: :identifier
```

> Will now use `params[:identifier]` instead of `params[:id]`.

Reviewer: ```suggestion
The generated routes will look like this now:
```

| PATCH/PUT | /kategorien/:id            | categories#update  | category_path(:id)      |
| DELETE    | /kategorien/:id            | categories#destroy | category_path(:id)      |

### Specifying the Singular Form of a Resource

> If you need to override the singular form of a resource, you can a rule to Active Support Inflector via [`inflections`][]:

Reviewer: ```suggestion
If you need to override the singular form of a resource, you can add a rule to Active Support Inflector via [`inflections`][]:
```

edit_video_path(video) # => "/videos/Roman-Holiday/edit"
```

Breaking Up a Large Route File With `draw`
-----------------------------------------

In a large application with thousands of routes, a single `config/routes.rb` file can become cumbersome and hard to read. Rails offers a way to break up a single `routes.rb` file into multiple small ones using the [`draw`][] macro.

> For example, you could add an `admin.rb` file that contains all the routes related to admin, another `api.rb` file for API related resources, etc.

Reviewer: ```suggestion
For example, you could add an `admin.rb` file that contains all the routes related to the admin area, another `api.rb` file for API-related resources, etc.
```

end
```

```irb
irb> video = Video.find_by(identifier: "Roman-Holiday")
irb> edit_video_path(video)
=> "/videos/Roman-Holiday/edit"
```

Breaking Up a Large Route File With `draw`

> -----------------------------------------

Reviewer 2: Breaking up large route files is only useful for very large applications (and even then we warn against using it). Inspecting and testing routes seems more common/important. Maybe move this section to the bottom?

edit_user GET    /users/:id/edit(.:format) users#edit
```

> The route name (`new_user` above for example) can be considered the base for deriving route helpers. For example to get the name of a route helper, add the suffix “_path” or “_url” to the route name(`new_user_path` for example).

Reviewer: [The rendered guides](https://ce08d9e7.rails-docs-preview.pages.dev/guides/routing#inspecting-routes) is recognizing the underscores as italics. I'd say we should use a code block here:

```suggestion
The route name (`new_user` above, for example) can be considered the base for deriving route helpers. To get the name of a route helper, add the suffix `_path` or `_url` to the route name(`new_user_path`, for example).
```

Controller#Action | users#edit
```

### Routes in Rails Console

> You can access route helpers using `Rails.application.routes.url_helpers` within the [Rails Console](command_line.html#bin-rails-console). For example:
> 
> ```irb
> irb> app.users_path
> => "/users"
> 
> irb> user = User.first
> => #<User:0x00007fc1eab81628
> irb> app.edit_user_path(user)
> => "/users/1/edit"
> ```

Reviewer: I find it confusing to state that you can use helpers via `` when the example uses `app.url_helper` directly.

```suggestion
You can access route helpers using `Rails.application.routes.url_helpers` within the [Rails Console](command_line.html#bin-rails-console). They are also available via the [app](command_line.html#the-app-and-helper-objects) object. For example:

```irb
irb> Rails.application.routes.url_helpers.users_path
=> "/users"

irb> user = User.first
=> #<User:0x00007fc1eab81628
irb> app.edit_user_path(user)
=> "/users/1/edit"
```
```

Author: That's a good clarification, thanks!

Controller#Action | users#edit
```

> ### Routes in Rails Console

Reviewer 2: The previous section mentioned `bin/rails routes --expanded`.
The next section mentions `bin/rails routes -g new_comment`.
After that `bin/rails routes --unused`.
Maybe move this console section after all `bin/rails routes` sections?

Author: makes sense

$ bin/rails routes -c Articles::CommentsController
```

TIP: The output from `bin/rails routes` is easier to read if you widen your terminal window until the output lines don't wrap.

> ### Listing Unused Routes

Reviewer 5: This is a great new section. A sentence or two on what unused routes mean would be useful. My suggestion:

An "unused" route in Rails is a route that is defined in the `config/routes.rb` file but is not referenced by any controller action in your application.

The `--unused` option is useful for cleaning up your routes file, identifying potential dead code, or understanding which parts of your routing configuration might not be actively used in your application.

NOTE: An "unused" route doesn't necessarily mean it's not needed - it might be planned for future use or used in a way that Rails doesn't automatically detect. Review the results carefully before removing any routes.

Reviewer: A suggestion on top of @hrishimittal 's:

```
An "unused" route in Rails is a route that is defined in the config/routes.rb file but is not referenced by any controller action or view in your application.
```

You can have a route defined directly to a view without an explicit controller action.

$ bin/rails routes -c Articles::CommentsController
```

> TIP: The output from `bin/rails routes` is easier to read if you widen your terminal window until the output lines don't wrap.

Reviewer: We could mention that the `--expanded` option is useful in this case too

$ bin/rails routes -c Articles::CommentsController
```

TIP: The output from `bin/rails routes` is easier to read if you widen your terminal window until the output lines don't wrap or use the `--expanded` option.

### Listing Unused Routes

You can scan your application for unused routes with the `--unused` option. An "unused" route in Rails is a route that is defined in the config/routes.rb file but is not referenced by any controller action or view in your application. For example:

```bash

> $ rails routes --unused

Reviewer 2: ```suggestion
$ bin/rails routes --unused
```
